In [1]:
import pandas as pd

In [2]:
# url='https://raw.githubusercontent.com/pandalearnstocode/csvfiles/master/salesdata.csv'
# salesdata=pd.read_csv('/Users/shubhdeepsarkar/Developer/datasets/salesdata.csv',parse_dates=['DATES'])

In [3]:
df_athlete_events = pd.read_csv('/Users/shubhdeepsarkar/Developer/datasets/bi_pt_archive/athlete_events.csv')
df_noc_region = df = pd.read_csv('/Users/shubhdeepsarkar/Developer/datasets/bi_pt_archive/noc_regions.csv')

# Taking a look at datasets

In [4]:
df_athlete_events.head()

ID                      Name Sex   Age  Height  Weight            Team  \
0   1                 A Dijiang   M  24.0   180.0    80.0           China   
1   2                  A Lamusi   M  23.0   170.0    60.0           China   
2   3       Gunnar Nielsen Aaby   M  24.0     NaN     NaN         Denmark   
3   4      Edgar Lindenau Aabye   M  34.0     NaN     NaN  Denmark/Sweden   
4   5  Christine Jacoba Aaftink   F  21.0   185.0    82.0     Netherlands   

   NOC        Games  Year  Season       City          Sport  \
0  CHN  1992 Summer  1992  Summer  Barcelona     Basketball   
1  CHN  2012 Summer  2012  Summer     London           Judo   
2  DEN  1920 Summer  1920  Summer  Antwerpen       Football   
3  DEN  1900 Summer  1900  Summer      Paris     Tug-Of-War   
4  NED  1988 Winter  1988  Winter    Calgary  Speed Skating   

                              Event Medal  
0       Basketball Men's Basketball   NaN  
1      Judo Men's Extra-Lightweight   NaN  
2           Football Men's Football   NaN  
3       Tug-Of-War Men's Tug-Of-War  Gold  
4  Speed Skating Women's 500 metres   NaN

In [5]:
df_noc_region.head()

NOC       region                 notes
0  AFG  Afghanistan                   NaN
1  AHO      Curacao  Netherlands Antilles
2  ALB      Albania                   NaN
3  ALG      Algeria                   NaN
4  AND      Andorra                   NaN

In [6]:
df_athlete_events.shape

(271116, 15)

In [7]:
df_noc_region.shape

(230, 3)

In [8]:
df_athlete_events.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 271116 entries, 0 to 271115
Data columns (total 15 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   ID      271116 non-null  int64  
 1   Name    271116 non-null  object 
 2   Sex     271116 non-null  object 
 3   Age     261642 non-null  float64
 4   Height  210945 non-null  float64
 5   Weight  208241 non-null  float64
 6   Team    271116 non-null  object 
 7   NOC     271116 non-null  object 
 8   Games   271116 non-null  object 
 9   Year    271116 non-null  int64  
 10  Season  271116 non-null  object 
 11  City    271116 non-null  object 
 12  Sport   271116 non-null  object 
 13  Event   271116 non-null  object 
 14  Medal   39783 non-null   object 
dtypes: float64(3), int64(2), object(10)
memory usage: 31.0+ MB


In [9]:
df_noc_region.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 230 entries, 0 to 229
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   NOC     230 non-null    object
 1   region  227 non-null    object
 2   notes   21 non-null     object
dtypes: object(3)
memory usage: 5.5+ KB


# Cleaning the datasets

## Dropping unnecessary columns

In [11]:
df_noc_region.drop('notes',axis=1, inplace=True)

In [12]:
df_noc_region.head()

NOC       region
0  AFG  Afghanistan
1  AHO      Curacao
2  ALB      Albania
3  ALG      Algeria
4  AND      Andorra

## Merging regions and athlete datasets

In [21]:
df = pd.merge(df_athlete_events, df_noc_region, on=['NOC'], how='left')

In [23]:
df.head()

ID                      Name Sex   Age  Height  Weight            Team  \
0   1                 A Dijiang   M  24.0   180.0    80.0           China   
1   2                  A Lamusi   M  23.0   170.0    60.0           China   
2   3       Gunnar Nielsen Aaby   M  24.0     NaN     NaN         Denmark   
3   4      Edgar Lindenau Aabye   M  34.0     NaN     NaN  Denmark/Sweden   
4   5  Christine Jacoba Aaftink   F  21.0   185.0    82.0     Netherlands   

   NOC        Games  Year  Season       City          Sport  \
0  CHN  1992 Summer  1992  Summer  Barcelona     Basketball   
1  CHN  2012 Summer  2012  Summer     London           Judo   
2  DEN  1920 Summer  1920  Summer  Antwerpen       Football   
3  DEN  1900 Summer  1900  Summer      Paris     Tug-Of-War   
4  NED  1988 Winter  1988  Winter    Calgary  Speed Skating   

                              Event Medal       region  
0       Basketball Men's Basketball   NaN        China  
1      Judo Men's Extra-Lightweight   NaN        China  
2           Football Men's Football   NaN      Denmark  
3       Tug-Of-War Men's Tug-Of-War  Gold      Denmark  
4  Speed Skating Women's 500 metres   NaN  Netherlands

## Removing duplicates

In [24]:
df.duplicated().sum()

1385

In [25]:
df[df.duplicated()].head()

ID                    Name Sex   Age  Height  Weight           Team  \
1252   704      Dsir Antoine Acket   M  27.0     NaN     NaN        Belgium   
4282  2449  William Truman Aldrich   M  48.0     NaN     NaN  United States   
4283  2449  William Truman Aldrich   M  48.0     NaN     NaN  United States   
4862  2777  Hermann Reinhard Alker   M  43.0     NaN     NaN        Germany   
4864  2777  Hermann Reinhard Alker   M  43.0     NaN     NaN        Germany   

      NOC        Games  Year  Season         City             Sport  \
1252  BEL  1932 Summer  1932  Summer  Los Angeles  Art Competitions   
4282  USA  1928 Summer  1928  Summer    Amsterdam  Art Competitions   
4283  USA  1928 Summer  1928  Summer    Amsterdam  Art Competitions   
4862  GER  1928 Summer  1928  Summer    Amsterdam  Art Competitions   
4864  GER  1928 Summer  1928  Summer    Amsterdam  Art Competitions   

                                                  Event Medal   region  
1252     Art Competitions Mixed Painting, Unknown Event   NaN  Belgium  
4282  Art Competitions Mixed Painting, Drawings And ...   NaN      USA  
4283  Art Competitions Mixed Painting, Drawings And ...   NaN      USA  
4862  Art Competitions Mixed Architecture, Designs F...   NaN  Germany  
4864  Art Competitions Mixed Architecture, Architect...   NaN  Germany

In [26]:
df.drop_duplicates(inplace=True)

In [28]:
df.duplicated().sum()

0

## Null values treatment

In [45]:
df.isnull().sum()

ID        0
Name      0
Sex       0
Age       0
Height    0
Weight    0
Team      0
NOC       0
Games     0
Year      0
Season    0
City      0
Sport     0
Event     0
Medal     0
region    0
dtype: int64

In [31]:
df.fillna('0', inplace=True)

In [46]:
df.isnull().sum()

ID        0
Name      0
Sex       0
Age       0
Height    0
Weight    0
Team      0
NOC       0
Games     0
Year      0
Season    0
City      0
Sport     0
Event     0
Medal     0
region    0
dtype: int64

## Changing datatypes of height, weight, and age

In [37]:
df.Age = df.Age.astype(int)
df.Height = df.Height.astype(float)
df.Weight = df.Weight.astype(float)


In [39]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 269731 entries, 0 to 271115
Data columns (total 16 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   ID      269731 non-null  int64  
 1   Name    269731 non-null  object 
 2   Sex     269731 non-null  object 
 3   Age     269731 non-null  int64  
 4   Height  269731 non-null  float64
 5   Weight  269731 non-null  float64
 6   Team    269731 non-null  object 
 7   NOC     269731 non-null  object 
 8   Games   269731 non-null  object 
 9   Year    269731 non-null  int64  
 10  Season  269731 non-null  object 
 11  City    269731 non-null  object 
 12  Sport   269731 non-null  object 
 13  Event   269731 non-null  object 
 14  Medal   269731 non-null  object 
 15  region  269731 non-null  object 
dtypes: float64(2), int64(3), object(11)
memory usage: 35.0+ MB


*Some games were removed from Olympics. Like Cricket, Roquet.* <br>
*Removed sports records do not affect the overall dataset and they are in a fraction of quantity, hence there is no sense in dropping them.* <br>
*Moreover, some useful insights can be made only in there presence.*

---
<br>

# EDA

### 1. How many olympics games have been held?

In [194]:
df.Games.nunique()

51

### 2. List down all Olympics games held so far.

In [48]:
pd.DataFrame({'Games': sorted(list(df.Games.unique()))})

Games
0   1896 Summer
1   1900 Summer
2   1904 Summer
3   1906 Summer
4   1908 Summer
5   1912 Summer
6   1920 Summer
7   1924 Summer
8   1924 Winter
9   1928 Summer
10  1928 Winter
11  1932 Summer
12  1932 Winter
13  1936 Summer
14  1936 Winter
15  1948 Summer
16  1948 Winter
17  1952 Summer
18  1952 Winter
19  1956 Summer
20  1956 Winter
21  1960 Summer
22  1960 Winter
23  1964 Summer
24  1964 Winter
25  1968 Summer
26  1968 Winter
27  1972 Summer
28  1972 Winter
29  1976 Summer
30  1976 Winter
31  1980 Summer
32  1980 Winter
33  1984 Summer
34  1984 Winter
35  1988 Summer
36  1988 Winter
37  1992 Summer
38  1992 Winter
39  1994 Winter
40  1996 Summer
41  1998 Winter
42  2000 Summer
43  2002 Winter
44  2004 Summer
45  2006 Winter
46  2008 Summer
47  2010 Winter
48  2012 Summer
49  2014 Winter
50  2016 Summer

### 3. Mention the total no of nations who participated in each olympics game?

In [49]:
df[['Games','NOC']].groupby(['Games']).nunique()

NOC
Games           
1896 Summer   12
1900 Summer   31
1904 Summer   15
1906 Summer   21
1908 Summer   22
1912 Summer   29
1920 Summer   29
1924 Summer   45
1924 Winter   19
1928 Summer   46
1928 Winter   25
1932 Summer   47
1932 Winter   17
1936 Summer   49
1936 Winter   28
1948 Summer   59
1948 Winter   28
1952 Summer   69
1952 Winter   30
1956 Summer   72
1956 Winter   32
1960 Summer   84
1960 Winter   30
1964 Summer   93
1964 Winter   36
1968 Summer  112
1968 Winter   37
1972 Summer  121
1972 Winter   35
1976 Summer   92
1976 Winter   37
1980 Summer   80
1980 Winter   37
1984 Summer  140
1984 Winter   49
1988 Summer  159
1988 Winter   57
1992 Summer  169
1992 Winter   64
1994 Winter   67
1996 Summer  197
1998 Winter   72
2000 Summer  200
2002 Winter   77
2004 Summer  201
2006 Winter   79
2008 Summer  204
2010 Winter   82
2012 Summer  205
2014 Winter   89
2016 Summer  207

### 4. Which year saw the highest and lowest no of countries participating in olympics?

In [50]:
df_year_NOC = pd.DataFrame({'Year': sorted(list(df.Year.unique())), 'NOCs': df.groupby(['Year']).nunique().NOC}, )

In [51]:
min_ = df_year_NOC.NOCs.min()
min_year = list(df_year_NOC.Year[df_year_NOC.NOCs== df_year_NOC.NOCs.min()])[0]

max_ = df_year_NOC.NOCs.max()
max_year = list(df_year_NOC.Year[df_year_NOC.NOCs== df_year_NOC.NOCs.max()])[0]

In [52]:
print(f"Year {min_year} had the least number of participating countries, {min_}")
print(f"Year {max_year} saw the maximum number of countries participating, {max_}")


Year 1896 had the least number of participating countries, 12
Year 2016 saw the maximum number of countries participating, 207


### 5. Which nation has participated in all of the olympic games?

In [195]:
every_time = list(df.groupby(['region']).nunique()[df.groupby(['region']).nunique().Games == 51].index)
for _ in every_time:
    print()

France
Italy
Switzerland
UK


### 6. Identify the sport which was played in all summer olympics.

In [56]:
df.groupby('Season').nunique().Games

Season
Summer    29
Winter    22
Name: Games, dtype: int64

So there are a total of 29 Summer Olympics till 2016

In [57]:
df_sport = df[df.Season == 'Summer'].groupby('Sport').nunique()

In [196]:
every_time_sports = list(df_sport.Games[df_sport.Games == 29].index)
for _ in every_time_sports:
    print(_)

Athletics
Cycling
Fencing
Gymnastics
Swimming


### 7. Which Sports were just played only once in the olympics?

In [60]:
pd.DataFrame(list(df.groupby('Sport').nunique()[df.groupby('Sport').nunique().Games == 1].index))

0
0          Aeronautics
1        Basque Pelota
2              Cricket
3              Croquet
4         Jeu De Paume
5  Military Ski Patrol
6         Motorboating
7             Racquets
8                Roque
9         Rugby Sevens

### 8. Fetch the total no of sports played in each olympic games.

In [61]:
df[['Games', 'Sport']].groupby('Games').nunique()

Sport
Games             
1896 Summer      9
1900 Summer     20
1904 Summer     18
1906 Summer     13
1908 Summer     24
1912 Summer     17
1920 Summer     25
1924 Summer     20
1924 Winter     10
1928 Summer     17
1928 Winter      8
1932 Summer     18
1932 Winter      7
1936 Summer     24
1936 Winter      8
1948 Summer     20
1948 Winter      9
1952 Summer     19
1952 Winter      8
1956 Summer     19
1956 Winter      8
1960 Summer     19
1960 Winter      8
1964 Summer     21
1964 Winter     10
1968 Summer     20
1968 Winter     10
1972 Summer     23
1972 Winter     10
1976 Summer     23
1976 Winter     10
1980 Summer     23
1980 Winter     10
1984 Summer     25
1984 Winter     10
1988 Summer     27
1988 Winter     10
1992 Summer     29
1992 Winter     12
1994 Winter     12
1996 Summer     31
1998 Winter     14
2000 Summer     34
2002 Winter     15
2004 Summer     34
2006 Winter     15
2008 Summer     34
2010 Winter     15
2012 Summer     32
2014 Winter     15
2016 Summer     34

### 9. Fetch details of the oldest athletes to win a gold medal.

In [62]:
df[df.Medal == 'Gold'].sort_values('Age', ascending=False).head()

ID                          Name Sex  Age  Height  Weight  \
233390  117046             Oscar Gomer Swahn   M   64     0.0     0.0   
105199   53238               Charles Jacobus   M   64     0.0     0.0   
261102  130662       Robert W. Williams, Jr.   M   63     0.0     0.0   
104003   52626          Isaac Lazarus Israls   M   63     0.0     0.0   
226374  113773  Galen Carter "G. C." Spencer   M   63   165.0     0.0   

                   Team  NOC        Games  Year  Season       City  \
233390           Sweden  SWE  1912 Summer  1912  Summer  Stockholm   
105199    United States  USA  1904 Summer  1904  Summer  St. Louis   
261102  Potomac Archers  USA  1904 Summer  1904  Summer  St. Louis   
104003      Netherlands  NED  1928 Summer  1928  Summer  Amsterdam   
226374  Potomac Archers  USA  1904 Summer  1904  Summer  St. Louis   

                   Sport                                             Event  \
233390          Shooting  Shooting Men's Running Target, Single Shot, Team   
105199             Roque                               Roque Men's Singles   
261102           Archery                          Archery Men's Team Round   
104003  Art Competitions        Art Competitions Mixed Painting, Paintings   
226374           Archery                          Archery Men's Team Round   

       Medal       region  
233390  Gold       Sweden  
105199  Gold          USA  
261102  Gold          USA  
104003  Gold  Netherlands  
226374  Gold          USA

### 10. Find the Ratio of male and female athletes participated in all olympic games.

In [230]:
df_male_female = pd.pivot_table(df.groupby(['Games', 'Sex']).count(), index='Games', columns='Sex')['Name']
df_male_female.fillna(0, inplace=True)
female_percent = round((df_male_female.F/(df_male_female.M + df_male_female.F)) * 100, 2)
male_percent = round((df_male_female.M/(df_male_female.M + df_male_female.F)) * 100, 2)

In [228]:
df_male_female['Female %'] = female_percent
df_male_female['Male %'] = male_percent

In [229]:
df_male_female

Sex               F       M  Female %  Male %
Games                                        
1896 Summer     0.0   380.0      0.00  100.00
1900 Summer    33.0  1865.0      1.74   98.26
1904 Summer    16.0  1285.0      1.23   98.77
1906 Summer    11.0  1722.0      0.63   99.37
1908 Summer    47.0  3022.0      1.53   98.47
1912 Summer    87.0  3953.0      2.15   97.85
1920 Summer   134.0  4158.0      3.12   96.88
1924 Summer   234.0  4876.0      4.58   95.42
1924 Winter    17.0   443.0      3.70   96.30
1928 Summer   391.0  4265.0      8.40   91.60
1928 Winter    33.0   549.0      5.67   94.33
1932 Summer   265.0  2200.0     10.75   89.25
1932 Winter    22.0   330.0      6.25   93.75
1936 Summer   447.0  5804.0      7.15   92.85
1936 Winter    81.0   814.0      9.05   90.95
1948 Summer   610.0  5698.0      9.67   90.33
1948 Winter   133.0   942.0     12.37   87.63
1952 Summer  1497.0  6773.0     18.10   81.90
1952 Winter   185.0   903.0     17.00   83.00
1956 Summer   893.0  4234.0     17.42   82.58
1956 Winter   246.0  1061.0     18.82   81.18
1960 Summer  1435.0  6684.0     17.67   82.33
1960 Winter   295.0   821.0     26.43   73.57
1964 Summer  1348.0  6354.0     17.50   82.50
1964 Winter   404.0  1374.0     22.72   77.28
1968 Summer  1777.0  6811.0     20.69   79.31
1968 Winter   416.0  1475.0     22.00   78.00
1972 Summer  2193.0  8111.0     21.28   78.72
1972 Winter   415.0  1240.0     25.08   74.92
1976 Summer  2172.0  6469.0     25.14   74.86
1976 Winter   434.0  1427.0     23.32   76.68
1980 Summer  1756.0  5435.0     24.42   75.58
1980 Winter   430.0  1316.0     24.63   75.37
1984 Summer  2447.0  7007.0     25.88   74.12
1984 Winter   536.0  1598.0     25.12   74.88
1988 Summer  3543.0  8494.0     29.43   70.57
1988 Winter   680.0  1959.0     25.77   74.23
1992 Summer  4124.0  8853.0     31.78   68.22
1992 Winter  1054.0  2382.0     30.68   69.32
1994 Winter  1105.0  2055.0     34.97   65.03
1996 Summer  5008.0  8772.0     36.34   63.66
1998 Winter  1384.0  2221.0     38.39   61.61
2000 Summer  5431.0  8390.0     39.30   60.70
2002 Winter  1582.0  2527.0     38.50   61.50
2004 Summer  5546.0  7897.0     41.26   58.74
2006 Winter  1757.0  2625.0     40.10   59.90
2008 Summer  5816.0  7786.0     42.76   57.24
2010 Winter  1847.0  2555.0     41.96   58.04
2012 Summer  5815.0  7105.0     45.01   54.99
2014 Winter  2023.0  2868.0     41.36   58.64
2016 Summer  6223.0  7465.0     45.46   54.54

### 11. Fetch the top 5 athletes who have won the most gold medals.

In [83]:
pd.DataFrame(df[df.Medal == 'Gold'].groupby('Name').count().sort_values('Medal', ascending=False).Medal.head())

Medal
Name                                     
Michael Fred Phelps, II                23
Raymond Clarence "Ray" Ewry            10
Mark Andrew Spitz                       9
Paavo Johannes Nurmi                    9
Larysa Semenivna Latynina (Diriy-)      9

### 12. Fetch the top 5 athletes who have won the most medals (gold/silver/bronze).

In [86]:
pd.DataFrame(df.groupby('Name').count().sort_values('Medal', ascending=False).Medal.head())

Medal
Name                           
Heikki Ilmari Savolainen     39
Joseph "Josy" Stoffel        38
Ioannis Theofilakis          36
Takashi Ono                  33
Andreas Wecker               32

### 13. Fetch the top 5 most successful countries in olympics. Success is defined by no of medals won.

In [90]:
df[['region', 'Medal']][df.Medal.notnull()].groupby('region').count().sort_values('Medal', ascending=False).head()


Medal
region        
USA      18604
Germany  15787
France   12551
UK       12115
Russia   11692

<!-- ![image.png](attachment:image.png) -->

![ChessUrl](https://media0.giphy.com/media/oCjCwnuLpiWbfMb1UA/giphy.gif "chess")


### 14. List down total gold, silver and broze medals won by each country.

In [145]:
import numpy as np

pivot_medals = pd.pivot_table(df, index='region', columns='Medal', aggfunc='count')
pivot_medals['Games'][['Gold', 'Silver', 'Bronze']].fillna(0).sort_values(['Gold', 'Silver', 'Bronze'], ascending=False).astype(int)

Medal                    Gold  Silver  Bronze
region                                       
USA                      2638    1641    1358
Russia                   1599    1170    1178
Germany                  1301    1195    1260
UK                        677     739     651
Italy                     575     531     531
...                       ...     ...     ...
Timor-Leste                 0       0       0
Turkmenistan                0       0       0
Vanuatu                     0       0       0
Virgin Islands, British     0       0       0
Yemen                       0       0       0

[206 rows x 3 columns]

### 15. List down total gold, silver and broze medals won by each country corresponding to each olympic games.

In [146]:
pivot_medals0 = pd.pivot_table(df, index=['region', 'Games'], columns='Medal', aggfunc='count')
pivot_medals0['Year'][['Gold', 'Silver', 'Bronze']].fillna(0).sort_values(['region']).astype(int)

Medal                 Gold  Silver  Bronze
region   Games                            
0        1912 Summer     0       0       0
         2016 Summer     1       0       0
         2012 Summer     0       0       4
         2008 Summer     0       3       0
         2004 Summer     0       0       0
...                    ...     ...     ...
Zimbabwe 1960 Summer     0       0       0
         1928 Summer     0       0       0
         2014 Winter     0       0       0
         1988 Summer     0       0       0
         2016 Summer     0       0       0

[3803 rows x 3 columns]

### 16. Identify which country won the most gold, most silver and most bronze medals in each olympic games.

In [148]:
pd.pivot_table(df, index='Games', columns='Medal', aggfunc='max')['region'][['Gold', 'Silver', 'Bronze']]

Medal              Gold              Silver       Bronze
Games                                                   
1896 Summer         USA                 USA          USA
1900 Summer         USA                 USA          USA
1904 Summer         USA                 USA          USA
1906 Summer         USA                 USA          USA
1908 Summer         USA                 USA          USA
1912 Summer         USA                 USA          USA
1920 Summer         USA                 USA          USA
1924 Summer     Uruguay                 USA          USA
1924 Winter         USA                 USA          USA
1928 Summer     Uruguay                 USA          USA
1928 Winter         USA                 USA          USA
1932 Summer         USA                 USA      Uruguay
1932 Winter         USA                 USA          USA
1936 Summer         USA                 USA          USA
1936 Winter         USA                  UK          USA
1948 Summer         USA             Uruguay      Uruguay
1948 Winter         USA                 USA          USA
1952 Summer         USA                 USA    Venezuela
1952 Winter         USA                 USA          USA
1956 Summer         USA                 USA      Uruguay
1956 Winter         USA                 USA          USA
1960 Summer         USA                 USA    Venezuela
1960 Winter         USA                 USA          USA
1964 Summer         USA                 USA      Uruguay
1964 Winter         USA                 USA          USA
1968 Summer   Venezuela              Uganda       Uganda
1968 Winter         USA                 USA          USA
1972 Summer      Uganda              Uganda          USA
1972 Winter         USA                 USA          USA
1976 Summer         USA           Venezuela          USA
1976 Winter         USA                 USA          USA
1980 Summer    Zimbabwe           Venezuela           UK
1980 Winter         USA                 USA          USA
1984 Summer         USA                 USA       Zambia
1984 Winter         USA                 USA  Switzerland
1988 Summer         USA  Virgin Islands, US          USA
1988 Winter         USA                 USA          USA
1992 Summer         USA                 USA          USA
1992 Winter         USA                 USA          USA
1994 Winter  Uzbekistan                 USA      Ukraine
1996 Summer     Ukraine              Zambia   Uzbekistan
1998 Winter         USA             Ukraine          USA
2000 Summer  Uzbekistan             Vietnam   Uzbekistan
2002 Winter         USA                 USA          USA
2004 Summer    Zimbabwe            Zimbabwe     Zimbabwe
2006 Winter         USA                 USA      Ukraine
2008 Summer    Zimbabwe            Zimbabwe    Venezuela
2010 Winter         USA                 USA          USA
2012 Summer   Venezuela             Ukraine   Uzbekistan
2014 Winter     Ukraine                 USA      Ukraine
2016 Summer     Vietnam             Vietnam    Venezuela

### 17. Identify which country won the most gold, most silver, most bronze medals and the most medals in each olympic games.

In [180]:
df_region = pd.pivot_table(df, index='Games', columns='Medal', aggfunc='max')['region'][['Gold', 'Silver', 'Bronze']]

df_medal_count = pd.pivot_table(df, index='Games', columns='Medal', aggfunc='count')['region'][['Gold', 'Silver', 'Bronze']]

gold = df_region.Gold + ', ' +list(map(str, (df_medal_count.Gold)))
silver = df_region.Silver + ', ' +list(map(str, (df_medal_count.Silver)))
bronze = df_region.Bronze + ', ' +list(map(str, (df_medal_count.Bronze)))

df_most_medal_country = pd.DataFrame(index=list(df_region.index))



In [184]:
df_most_medal_country['Most_gold'] = list(gold)
df_most_medal_country['Most_silver'] = list(silver)
df_most_medal_country['Most_bronze'] = list(bronze)

In [185]:
df_most_medal_country

Most_gold              Most_silver      Most_bronze
1896 Summer          USA, 62                  USA, 43          USA, 38
1900 Summer         USA, 198                 USA, 220         USA, 175
1904 Summer         USA, 173                 USA, 163         USA, 150
1906 Summer         USA, 157                 USA, 156         USA, 145
1908 Summer         USA, 294                 USA, 281         USA, 256
1912 Summer         USA, 326                 USA, 315         USA, 300
1920 Summer         USA, 493                 USA, 448         USA, 367
1924 Summer     Uruguay, 277                 USA, 281         USA, 274
1924 Winter          USA, 55                  USA, 38          USA, 37
1928 Summer     Uruguay, 245                 USA, 239         USA, 250
1928 Winter          USA, 30                  USA, 28          USA, 31
1932 Summer         USA, 229                 USA, 214     Uruguay, 204
1932 Winter          USA, 32                  USA, 32          USA, 28
1936 Summer         USA, 312                 USA, 310         USA, 295
1936 Winter          USA, 36                   UK, 37          USA, 35
1948 Summer         USA, 289             Uruguay, 284     Uruguay, 279
1948 Winter          USA, 41                  USA, 48          USA, 46
1952 Summer         USA, 306                 USA, 291   Venezuela, 300
1952 Winter          USA, 45                  USA, 44          USA, 47
1956 Summer         USA, 302                 USA, 293     Uruguay, 298
1956 Winter          USA, 51                  USA, 49          USA, 50
1960 Summer         USA, 309                 USA, 294   Venezuela, 308
1960 Winter          USA, 50                  USA, 48          USA, 49
1964 Summer         USA, 347                 USA, 339     Uruguay, 343
1964 Winter          USA, 61                  USA, 67          USA, 58
1968 Summer   Venezuela, 359              Uganda, 340      Uganda, 358
1968 Winter          USA, 66                  USA, 70          USA, 63
1972 Summer      Uganda, 404              Uganda, 392         USA, 419
1972 Winter          USA, 70                  USA, 63          USA, 66
1976 Summer         USA, 438           Venezuela, 434         USA, 448
1976 Winter          USA, 70                  USA, 71          USA, 70
1980 Summer    Zimbabwe, 457           Venezuela, 458          UK, 469
1980 Winter          USA, 72                  USA, 73          USA, 73
1984 Summer         USA, 497                 USA, 477      Zambia, 502
1984 Winter          USA, 74                  USA, 74  Switzerland, 74
1988 Summer         USA, 520  Virgin Islands, US, 513         USA, 549
1988 Winter          USA, 87                  USA, 88          USA, 88
1992 Summer         USA, 559                 USA, 549         USA, 604
1992 Winter         USA, 104                 USA, 108         USA, 106
1994 Winter  Uzbekistan, 110                 USA, 109     Ukraine, 112
1996 Summer     Ukraine, 608              Zambia, 605  Uzbekistan, 629
1998 Winter         USA, 145             Ukraine, 145         USA, 150
2000 Summer  Uzbekistan, 663             Vietnam, 661  Uzbekistan, 680
2002 Winter         USA, 162                 USA, 157         USA, 159
2004 Summer    Zimbabwe, 664            Zimbabwe, 660    Zimbabwe, 677
2006 Winter         USA, 176                 USA, 175     Ukraine, 175
2008 Summer    Zimbabwe, 671            Zimbabwe, 667   Venezuela, 710
2010 Winter         USA, 174                 USA, 175         USA, 171
2012 Summer   Venezuela, 632             Ukraine, 630  Uzbekistan, 679
2014 Winter     Ukraine, 202                 USA, 197     Ukraine, 198
2016 Summer     Vietnam, 665             Vietnam, 655   Venezuela, 703

### 18. Which countries have never won gold medal but have won silver/bronze medals?

In [110]:
import warnings
warnings.filterwarnings('ignore')

In [198]:
pivot_medals1 = pivot_medals['Year'][['Gold', 'Silver', 'Bronze']].fillna(0)

pivot_medals1.loc[(pivot_medals1.Gold == 0) & (pivot_medals1.Silver != 0) & (pivot_medals1.Bronze != 0)].sort_values(['Silver', 'Bronze'], ascending=False)

Medal         Gold  Silver  Bronze
region                            
Iceland        0.0    15.0     2.0
Malaysia       0.0    11.0     5.0
Philippines    0.0     3.0     7.0
Moldova        0.0     3.0     5.0
Lebanon        0.0     2.0     2.0
Ghana          0.0     1.0    22.0
Saudi Arabia   0.0     1.0     5.0
Qatar          0.0     1.0     4.0
Kyrgyzstan     0.0     1.0     2.0
Niger          0.0     1.0     1.0
Zambia         0.0     1.0     1.0

### 19. In which Sport/event, India has won highest medals.

In [113]:
pd.DataFrame(df.loc[(df.NOC == 'IND') & df.Medal.notnull()].groupby('Sport').count().sort_values('Medal').tail(1)['Medal'])

Medal
Sport        
Hockey    345

### 20. Break down all olympic games where India won medal for Hockey and how many medals in each olympic games.

In [117]:
pd.DataFrame(df.loc[(df.NOC == 'IND') & (df.Medal == 'Gold')].groupby('Games').count().sort_values('Medal', ascending=False)['Medal'])

Medal
Games             
1948 Summer     20
1936 Summer     19
1956 Summer     17
1980 Summer     16
1932 Summer     15
1964 Summer     15
1928 Summer     14
1952 Summer     14
1924 Winter      7
2008 Summer      1

---
<br>

## Conclusions/Insights

1. From 1896 to 2016, there are 51 Olympic games happened.
2. In 2016, a total of 207 countries participated in Olympics.
3. France, Italy, Switzerland, UK participated in all of the Olympics games.
4. Athletics, Cycling, Fencing, Gymnastics, Swimming were always present.
5. Aeronautics, Basque Pelota, Cricket and other 7 games had been removed from the Olympics.
6. A total of 34 games now played in Olympics and the least number of games played was just 8.
7. Oscar Gomer Swahn is oldest athlete (age: 64) ever to win a gold medal in shooting.
8. In 1896 Olympics there were no female athletes but it saw a steady increase over the years and eventually in 2016 Olympics female participants made a 45% of total athlete population.
9. Michael Fred Phelps, II won maximum Gold medals, 23 in Swimming
10. Heikki Ilmari Savolainen is higest medal scoring athlete (39) in Olympics.
11. USA won the most number of medals of all time, followed by Germany, France, UK and then Russia.
12. USA was the most active country scoring the most number of medals in almost every Olympic game since 1896.
13. Iceland, Malasiya, Philippines and 8 others country won Silver and Bronze, but never won any Gold medal. 
14. India won a total of 345 medals in Hockey.
15. India's performance in Hockey has declined over the years and they got only one 1 medal in 2008 Olympics.